core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java

```diff
@@ -1917,27 +1925,19 @@ public Locations locate(String tableName, Collection<Range> ranges)
       rangeList = new ArrayList<>(ranges);
     }
 
-    Map<String,Map<KeyExtent,List<Range>>> binnedRanges = new HashMap<>();
-
+    ClientTabletCache locator = ClientTabletCache.getInstance(context, tableId);
     locator.invalidateCache();
 
-    Retry retry = Retry.builder().infiniteRetries().retryAfter(100, MILLISECONDS)
-        .incrementBy(100, MILLISECONDS).maxWait(2, SECONDS).backOffFactor(1.5)
-        .logInterval(3, MINUTES).createRetry();
-
-    // ELASTICITY_TODO this will cause tablets to be hosted, but that may not be desired
-    while (!locator.binRanges(context, rangeList, binnedRanges).isEmpty()) {
-      context.requireTableExists(tableId, tableName);
-      context.requireNotOffline(tableId, tableName);
-      binnedRanges.clear();
-      try {
-        retry.waitForNextAttempt(log,
-            String.format("locating tablets in table %s(%s) for %d ranges", tableName, tableId,
-                rangeList.size()));
-      } catch (InterruptedException e) {
-        throw new RuntimeException(e);
-      }
-      locator.invalidateCache();
+    Map<String,Map<KeyExtent,List<Range>>> binnedRanges = new HashMap<>();
+    try {
+      @SuppressWarnings("unused")
+      List<Range> failed = locator.findTablets(context, rangeList, ((cachedTablet, range) -> {
+        if (cachedTablet.getTserverLocation().isPresent()) {
+          ClientTabletCacheImpl.addRange(binnedRanges, cachedTablet, range);
+        }
+      }), LocationNeed.NOT_REQUIRED);
```

Comment:
I locally tired making the method only work with the ALWAYS hosted goal.  Did not test this, it does compile. This is an alternative approach to deprecating and removing.  I like this approach.  If we go with this approach over deprecation then would need to update the javadoc to mention it only works with tablets that have a certain hosting goal.

```suggestion
Retry retry = Retry.builder().infiniteRetries().retryAfter(100, MILLISECONDS)
            .incrementBy(100, MILLISECONDS).maxWait(2, SECONDS).backOffFactor(1.5)
            .logInterval(3, MINUTES).createRetry();

    ArrayList<KeyExtent> locationLess = new ArrayList<>();
    Map<String,Map<KeyExtent,List<Range>>> binnedRanges = new HashMap<>();

    BiConsumer<CachedTablet,Range> rangeConsumer = (cachedTablet, range) -> {
      if (cachedTablet.getTserverLocation().isPresent()) {
        ClientTabletCacheImpl.addRange(binnedRanges, cachedTablet, range);
      } else {
        locationLess.add(cachedTablet.getExtent());
      }
    };

    try {

      List<Range> failed =
          locator.findTablets(context, rangeList, rangeConsumer, LocationNeed.NOT_REQUIRED);

      while (!failed.isEmpty() || !locationLess.isEmpty()) {
        context.requireTableExists(tableId, tableName);
        context.requireNotOffline(tableId, tableName);

        boolean allAlwaysHosted = context.getAmple().readTablets().forTablets(locationLess)
            .fetch(PREV_ROW, HOSTING_GOAL).build().stream()
            .allMatch(tmeta -> tmeta.getHostingGoal() == TabletHostingGoal.ALWAYS);

        if (!allAlwaysHosted) {
          throw new AccumuloException(
              "TableOperations.locate() only works with tablets that have a hosting goal of "
                  + TabletHostingGoal.ALWAYS
                  + ". Tablets with other hosting goals were seen.  table:" + tableName
                  + " table id:" + tableId);
        }

        try {
          retry.waitForNextAttempt(log,
              String.format("locating tablets in table %s(%s) for %d ranges", tableName, tableId,
                  rangeList.size()));
        } catch (InterruptedException e) {
          throw new RuntimeException(e);
        }

        locationLess.clear();
        binnedRanges.clear();
        locator.invalidateCache();
        failed = locator.findTablets(context, rangeList, rangeConsumer, LocationNeed.NOT_REQUIRED);
      }

```